OrbitLinuxTracing/PerfEventQueue.cpp

```diff
@@ -26,30 +23,76 @@ void PerfEventQueue::PushEvent(std::unique_ptr<PerfEvent> event) {
     const std::unique_ptr<std::queue<std::unique_ptr<PerfEvent>>>& queue = queue_it->second;
 
     queue->push(std::move(event));
-    queues_heap_.push(queue.get());
+    queues_heap_.emplace_back(queue.get());
+    MoveUpHeapBack();
   }
 }
 
 bool PerfEventQueue::HasEvent() { return !queues_heap_.empty(); }
 
-PerfEvent* PerfEventQueue::TopEvent() { return queues_heap_.top()->front().get(); }
+PerfEvent* PerfEventQueue::TopEvent() { return queues_heap_.front()->front().get(); }
 
 std::unique_ptr<PerfEvent> PerfEventQueue::PopEvent() {
-  std::queue<std::unique_ptr<PerfEvent>>* top_queue = queues_heap_.top();
-  queues_heap_.pop();
-
+  std::queue<std::unique_ptr<PerfEvent>>* top_queue = queues_heap_.front();
   std::unique_ptr<PerfEvent> top_event = std::move(top_queue->front());
   top_queue->pop();
+
   if (top_queue->empty()) {
     int top_fd = top_event->GetOriginFileDescriptor();
     queues_.erase(top_fd);
+    std::swap(queues_heap_.front(), queues_heap_.back());
+    queues_heap_.pop_back();
+    MoveDownHeapFront();
   } else {
-    // The queue with the older PerfEvent is always removed. If it's not empty, re-insert it so that
-    // it is in the right position in the heap after the front of the queue has been removed.
-    queues_heap_.push(top_queue);
+    MoveDownHeapFront();
   }
 
   return top_event;
 }
 
+void PerfEventQueue::MoveDownHeapFront() {
```

Comment:
To me those methods looks good. But with such algorithmic methods are always error-prune. I saw that you added an additional test case (thanks for that). Have you checked that the tests covers the corner cases here?

Reply:
Okay thanks. Yeah we never know if we have all the corner cases. I just wanted to double check. I guess thats totally fine now. Thanks!